source/tutorials/source-file-selection.md

```diff
@@ -0,0 +1,645 @@
+# Source file selection
+<!-- Note on title choice: While there's more uses outside of sources, it's by far the most prominent one -->
+
+<!-- TODO: Switch all mentions of unstable to stable once 23.11 is out -->
+
+To build a local project in a Nix derivation, its source files must be accessible to the builder.
+But since the builder runs in an isolated environment (if the [sandbox](https://nixos.org/manual/nix/stable/command-ref/conf-file.html#conf-sandbox) is enabled),
+it won't have access to the local project files by default.
+
+To make this work regardless, the Nix language has certain builtin features to copy local paths to the Nix store,
+whose paths are then accessible to derivation builders [^1].
+
+[^1]: Technically only Nix store paths from the derivations inputs can be accessed,
+but in practice this distinction is not important.
+
+These builtin features are very limited in functionality and are not recommended for anything non-trivial. For more advanced use cases, the file set library should be used instead.
+
+In this tutorial you'll learn both how to use the builtins and the file set library.
+
+## Setting up a local experiment
+
+To experiment with source file selection, we'll set up a local project.
+
+To start out, create a new directory, enter it, and set up `niv` to manage the Nixpkgs dependency:
+```shell-session
+$ mkdir select
+$ cd select
+$ nix-shell -p niv --run "niv init --nixpkgs nixos/nixpkgs --nixpkgs-branch nixos-unstable"
+```
+
+:::{note}
+For now we're using the nixos-unstable channel, since no stable channel has all the features we need yet.
+:::
+
+Then create a `default.nix` file with these contents:
+```nix
+{
+  system ? builtins.currentSystem,
+  sources ? import ./nix/sources.nix,
+}:
+let
+  pkgs = import sources.nixpkgs {
+    # Ensure purity
+    config = { };
+    overlays = [ ];
+    inherit system;
+  };
+in
+{
+  package = pkgs.callPackage ./package.nix { };
+  inherit pkgs;
+}
+```
+
+In this tutorial we'll experiment with different `package.nix` contents, while keeping `default.nix` the same.
+
+For now, let's have a simple `package.nix` to verify everything works so far:
+
+```nix
+{ runCommand }:
+runCommand "hello" { } ''
+  echo hello world
+''
+```
+
+And try it out:
+```shell-session
+$ nix-build -A package
+this derivation will be built:
+  /nix/store/kmf9sw8fn7ps3ndqs31hvqwsa35b8l3g-hello.drv
+building '/nix/store/kmf9sw8fn7ps3ndqs31hvqwsa35b8l3g-hello.drv'...
+hello world
+error: builder for '/nix/store/kmf9sw8fn7ps3ndqs31hvqwsa35b8l3g-hello.drv'
+  failed to produce output path for output 'out'
+```
+
+We could also add `touch $out` to make the build succeed,
+but we'll omit that for the sake of the tutorial, since we only need the build logs.
+This also makes it easier to build it again, since successful derivation builds would get cached.
+From now on we'll also make build outputs a bit shorter for the sake of brevity.
+
+## Builtins
+
+### Coercion of paths to strings
```

Comment:
@fricklerhandwerk I guess this section so far is specific to files, so maybe this should be reflected in the header.

Importing files into the store is explicitly not supported by `lib.fileset`, so this (or maybe even with `builtins.path` to set the store path name) is still the recommended way to do this.

Reply:
Actually this one could be removed as well. It's a related topic, but single files are generally not considered a "source".

With all of these sections removed, indeed this becomes very specific to file sets.